### PR TITLE
amd-ad: add type=sas network implementation

### PR DESCRIPTION
**Pull request by sthilt-sas**
Add the 'sas' ad type network implementation

### CODE REVIEW

**Comment by torch2424** [comment]
Thanks for adding your vendor to AMP!

Seems like you are failing some our travis tests. For the integration tests, seems like it could be a flake, but you definitely want to fix the issue with the unit test, which can be investigated with: `gulp check-types --local-changes`.

Once all of this is resolved, I can go ahead and do a review. Feel free to mention me once this is good to go. 

Thanks! 😄

**Comment by sthilt-sas** [comment]
@torch2424 Updated.  I fixed the issue related to my change. The failing tests now seem test timing related. Thanks!

File: build-system/conformance-config.textproto [hunk 1]
```diff
@@ -214,6 +214,7 @@ requirement: {
   whitelist: 'ads/kargo.js'
   whitelist: 'ads/mads.js'
   whitelist: 'ads/google/imaVideo.js'
+  whitelist: 'ads/sas.js'
```
[the remark this answers is not in the document]
Comment (zhouyx) [reply build-system/conformance-config.textproto]: LGTM

[the remark this answers is not in the document]
Comment (sthilt-sas) [reply build-system/conformance-config.textproto]: The error message (from 'conformance-config.textrpoto') says 'Use jsonParse instead. Usage in 3p ads can be whitelisted for now', which is why I added it to the whitelist

And 'src/json.js' is whitelisted in that group with a comment that says "Where jsonParse itself is implemented" -- but it's not. There's a parseJson method in src/json.js (note the verb-noun vs noun-verb).  'jsonParse' is defined multiple times in various node_modules.

In any case, I will un-whitelist my file and use parseJson from src/json.js as I see is done elsewhere.

**Comment by torch2424** [comment]
@sthilt-sas I see you made some of the requested changes. Thank you! 😄 

Feel free to mention me when this is ready for another round of review/testing. Thank you! 👍

**Comment by sthilt-sas** [comment]
@torch2424 Should be ready for review again. Thanks!

File: examples/ads.amp.html [hunk 2]
```diff
@@ -1602,6 +1603,18 @@ <h2>RUNative</h2>
       data-title="From the Web">
   </amp-embed>
 
+  <h2>SAS CI 360 Match</h2>
+  <amp-ad
+      width="468"
+      height="60"
+      type="sas"
```
Comment (torch2424) [inline examples/ads.amp.html]: Missing the 'layout' attribute 😄

Comment (sthilt-sas) [reply examples/ads.amp.html]: @torch2424 okay, added the layout attribute to the example and the documentation. Hopefully we all good now.

**Comment by torch2424** [comment]
So I tested this out, and I'm still getting: `Uncaught Error: Unknown ad type "sas"​​​`

screen shot 2019-02-13 at 1 27 38 pm

Though, comparing to other ad implementations, this should be working...

cc @zhouyx Could you take a look at this an see why?

**Comment by sthilt-sas** [comment]
> So I tested this out, and I'm still getting: `Uncaught Error: Unknown ad type "sas"​​​`

> cc @zhouyx Could you take a look at this an see why?

Yes, I saw this when I put type=sas in the url, but it displayed fine when displayed all ads. I also saw other newly added vendors showing the "unknown ad type" errors.

**Comment by torch2424** [comment]
@sthilt-sas So from my testing, I know I recently pulled in `type="tcsEmotion"`, and they work fine.

screen shot 2019-02-14 at 4 11 43 pm

But I finally got the ad to render with renderStart (perhaps it had something to do with the internet I was on at the office?)

screen shot 2019-02-14 at 4 17 04 pm

Will look through the code again, but will still defer to @zhouyx if we should merge this as is. As this is most likely an issue on our end.

**Comment by sthilt-sas** [comment]
@torch2424 okay tests are passing now

**Comment by zhouyx** [comment]
So we found the reason why the ad is not loaded with our example page. Our regex detects that there must be a whitespace after the amp-ad tagName, but you put an end of line instead. Because of this, the server cannot filter ad with type sas out. The fix is in #20899.

**Comment by sthilt-sas** [comment]
@torch2424 Are there any remaining issues with this? I fixed the travis linting issue and I think that was the last thing. Thanks!

**Comment by torch2424** [comment]
@sthilt-sas Nope, all issues are resolved.

Thank you very much for adding your vendor, and working with us on this! 😄 

And thank you as well @zhouyx for finding the regex bug.

Merging...

**Pull request merged**